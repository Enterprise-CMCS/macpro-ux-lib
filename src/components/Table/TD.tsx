import React, { PropsWithChildren } from "react";

type IntrinsicElements = JSX.IntrinsicElements["td"];

export interface TDProps extends IntrinsicElements {
  dataSortValue?: string;
}

/**
 *  **TD Component**
 *
 * @param {React.Node}  children       Contents to be rendered inside `<td>`.
 * @param {string}      dataSortValue  String value to be used when sorting the contents of this cell. If a dataSortValue is not provided, sort defaults to contents of the cell. (Ex: Dates should be converted to seconds or some other comparable value)
 */
export const TD: React.FC<PropsWithChildren<Props>> = ({
  children,
  dataSortValue,
  ...rest
}) => {
  return (
    <td data-sort-value={dataSortValue} {...rest}>
      {children}
    </td>
  );
};
