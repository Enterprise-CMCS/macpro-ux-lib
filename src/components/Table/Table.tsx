import React, { PropsWithChildren, useLayoutEffect, useRef } from "react";

import table from "@uswds/uswds/packages/usa-table/src";

type IntrinsicElements = JSX.IntrinsicElements["table"];

interface Props extends IntrinsicElements {
  borderless?: boolean;
  caption?: string;
  compact?: boolean;
  footnote?: string;
  scrollable?: boolean;
  stacked?: boolean;
  stackedHeader?: boolean;
  striped?: boolean;
}

/**
 * **Table Component**
 *
 * ```HTML
 * <Table>
 *   <thead>
 *     <tr>
 *       <TH>Header</TH>
 *     </tr>
 *   </thead>
 *   <tbody>
 *     <tr>
 *       <TD>Content</TD>
 *     </tr>
 *   </tbody>
 * </Table>
 * ```
 *
 * @param {boolean}     [borderless]      Removes the outer table borders, retaining only a single bottom border on each row. Best for tables with more text than numbers.
 * @param {string}      [caption]         Descriptive text information to be displayed above the table.
 * @param {React.Node}  [children]        Table content to be provided as children. Can use all standard HTML table elements (thead, tbody, tr, etc.). `TH` and `TD` are provided components that have some additional functionality.
 * @param {boolean}     [compact]         Reduces the row height and vertical spacing to display more table rows within a limited space. Should only be used with dense, numerical data, not text content. Pairs well with scrollable and striped variants, but is not suitable for use with stacked variants.
 * @param {string}      [footnote]        Descriptive text information to be displayed below the table.
 * @param {string}      id                Unique identifier required for the Table.
 * @param {boolean}     [scrollable]      Applies a horizontal scrollbar if the columns exceed the available width. Ideal for dense tables with many columns.
 * @param {boolean}     [stacked]         Stacks the table cells on narrow screens. Ideal for tables that contain more text information than numerical data. If you use this variant, you must ensure there is a data-label attribute on each cell of the table that matches the column header.
 * @param {boolean}     [stackedHeader]   Stacks the table cells on narrow screens and visually promotes the first cell of every row into a “header” for that group. Preferred for directories and other lists where the first cell of every row is a name. If you use this variant, you must ensure there is a data-label attribute on each cell of the table that matches the column header.
 * @param {boolean}     [striped]         Applies alternating horizontal striping to help the eye track across table rows. Pairs well with the scrollable variant for tables with many columns.
 */
export const Table: React.FC<PropsWithChildren<Props>> = ({
  borderless,
  caption,
  children,
  className,
  compact,
  footnote,
  id,
  scrollable,
  stacked,
  stackedHeader,
  striped,
  ...rest
}) => {
  const componentRef = useRef(null);

  useLayoutEffect(() => {
    if (typeof table.on === "function" && typeof table.off === "function") {
      const component = componentRef.current;
      table.off(component);
      table.on(component);
      return () => table.off(component);
    }
  }, [table]);

  const tableClassNames = `usa-table${
    borderless ? " usa-table--borderless" : ""
  }${compact ? " usa-table--compact" : ""}${
    stacked ? ` usa-table--stacked${stackedHeader ? "-header" : ""}` : ""
  }${striped ? " usa-table--striped" : ""}${className ? ` ${className}` : ""}`;

  return (
    <div
      ref={componentRef}
      className={`usa-table-container${
        scrollable ? " usa-table-container--scrollable" : ""
      }${stacked ? " width-mobile" : ""}`}
      tabIndex={0}
    >
      <table className={tableClassNames} {...rest}>
        {caption && <caption>{caption}</caption>}
        {children}
      </table>
      <div
        className="usa-sr-only usa-table__announcement-region"
        aria-live="polite"
      ></div>
      {footnote && (
        <p className="margin-bottom-3 usa-table__footnote">{footnote}</p>
      )}
    </div>
  );
};
