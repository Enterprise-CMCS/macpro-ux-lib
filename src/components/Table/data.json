{
  "scrollableTable": {
    "headers": [
      [
        {
          "value": "Federal Budget Baseline Projections",
          "props": {
            "rowSpan": 2
          }
        },
        {
          "value": "2017",
          "props": {
            "colSpan": 2
          }
        },
        {
          "value": "2018",
          "props": {
            "colSpan": 2
          }
        },
        {
          "value": "2019",
          "props": {
            "colSpan": 2
          }
        }
      ],
      [
        {
          "value": "%GDP"
        },
        {
          "value": "USD*"
        },
        {
          "value": "%GDP"
        },
        {
          "value": "USD*"
        },
        {
          "value": "%GDP"
        },
        {
          "value": "USD*"
        }
      ]
    ],
    "body": [
      [
        { "value": "Revenue" },
        { "value": "17.2%" },
        { "value": "3,316" },
        { "value": "16.4%" },
        { "value": "3,338" },
        { "value": "16.3%" },
        { "value": "3,490" }
      ],
      [
        { "value": "Outlays" },
        { "value": "20.6%" },
        { "value": "3,982" },
        { "value": "20.2%" },
        { "value": "4,142" },
        { "value": "21.0%" },
        { "value": "4,470" }
      ],
      [
        { "value": "Budget Deficit" },
        { "value": "-3.5%" },
        { "value": "-665" },
        { "value": "-3.8%" },
        { "value": "-804" },
        { "value": "-4.6%" },
        { "value": "-981" }
      ],
      [
        { "value": "Debt Held by Public" },
        { "value": "76.0%" },
        { "value": "14,665" },
        { "value": "77.4%" },
        { "value": "15,688" },
        { "value": "79.2%" },
        { "value": "16,762" }
      ]
    ]
  },
  "sortableTable": {
    "headers": [],
    "body": []
  },
  "standardTable": {
    "headers": [
      [
        {
          "value": "Document title"
        },
        {
          "value": "Description"
        },
        {
          "value": "Year"
        }
      ]
    ],
    "body": [
      [
        {
          "value": "Declaration of Independence"
        },
        {
          "value": "Statement adopted by the Continental Congress declaring independence from the British Empire."
        },
        {
          "value": "1776"
        }
      ],
      [
        {
          "value": "Bill of Rights"
        },
        {
          "value": "The first ten amendments of the U.S. Constitution guaranteeing rights and freedoms."
        },
        {
          "value": "1791"
        }
      ],
      [
        {
          "value": "Declaration of Sentiments"
        },
        {
          "value": "A document written during the Seneca Falls Convention outlining the rights that American women should be entitled to as citizens."
        },
        {
          "value": "1848"
        }
      ],
      [
        {
          "value": "Emancipation Proclamation"
        },
        {
          "value": "An executive order granting freedom to slaves in designated southern states. "
        },
        {
          "value": "1863"
        }
      ]
    ]
  }
}
