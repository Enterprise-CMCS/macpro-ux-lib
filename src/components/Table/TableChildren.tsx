import React from "react";
import { TD } from "./TD";
import { TH } from "./TH";

export const BorderlessTableChildren = [
  <thead key="BorderlessTableChildren__thead">
    <tr>
      <TH>Document title</TH>
      <TH>Description</TH>
      <TH>Year</TH>
    </tr>
  </thead>,
  <tbody key="BorderlessTableChildren__tbody">
    <tr>
      <TH rowHeader>Declaration of Independence</TH>
      <TD>
        Statement adopted by the Continental Congress declaring independence
        from the British Empire.
      </TD>
      <TD>1776</TD>
    </tr>
    <tr>
      <TH rowHeader>Bill of Rights</TH>
      <TD>
        The first ten amendments of the U.S. Constitution guaranteeing rights
        and freedoms.
      </TD>
      <TD>1791</TD>
    </tr>
    <tr>
      <TH rowHeader>Declaration of Sentiments</TH>
      <TD>
        A document written during the Seneca Falls Convention outlining the
        rights that American women should be entitled to as citizens.
      </TD>
      <TD>1848</TD>
    </tr>
    <tr>
      <TH rowHeader>Emancipation Proclamation</TH>
      <TD>
        An executive order granting freedom to slaves in designated southern
        states.
      </TD>
      <TD>1863</TD>
    </tr>
  </tbody>,
];

export const ScrollableTableChildren = [
  <colgroup key="ScrollableTableChildren__colgroup1">
    <col />
  </colgroup>,
  <colgroup span={2} key="ScrollableTableChildren__colgroup2"></colgroup>,
  <colgroup span={2} key="ScrollableTableChildren__colgroup3"></colgroup>,
  <thead key="ScrollableTableChildren__thead">
    <tr>
      <TH rowSpan={2}>
        Federal Budget
        <br />
        Baseline Projections
      </TH>
      <TH colSpan={2} scope="colgroup" className="text-center">
        2017
      </TH>
      <TH colSpan={2} scope="colgroup" className="text-center">
        2018
      </TH>
      <TH colSpan={2} scope="colgroup" className="text-center">
        2019
      </TH>
      <TH colSpan={2} scope="colgroup" className="text-center">
        2020
      </TH>
      <TH colSpan={2} scope="colgroup" className="text-center">
        2021
      </TH>
      <TH colSpan={2} scope="colgroup" className="text-center">
        Hist. Avg.
      </TH>
    </tr>
    <tr>
      <TH className="text-right">%GDP</TH>
      <TH className="text-right">USD*</TH>
      <TH className="text-right">%GDP</TH>
      <TH className="text-right">USD*</TH>
      <TH className="text-right">%GDP</TH>
      <TH className="text-right">USD*</TH>
      <TH className="text-right">%GDP</TH>
      <TH className="text-right">USD*</TH>
      <TH className="text-right">%GDP</TH>
      <TH className="text-right">USD*</TH>
      <TH className="text-right">%GDP</TH>
      <TH className="text-right">USD*</TH>
    </tr>
  </thead>,
  <tbody key="ScrollableTableChildren__tbody">
    <tr>
      <TH rowHeader>Revenue</TH>
      <TD className="font-mono-sm text-tabular text-right">17.2%</TD>
      <TD className="font-mono-sm text-tabular text-right">3,316</TD>
      <TD className="font-mono-sm text-tabular text-right">16.4%</TD>
      <TD className="font-mono-sm text-tabular text-right">3,338</TD>
      <TD className="font-mono-sm text-tabular text-right">16.3%</TD>
      <TD className="font-mono-sm text-tabular text-right">3,490</TD>
      <TD className="font-mono-sm text-tabular text-right">16.7%</TD>
      <TD className="font-mono-sm text-tabular text-right">3,678</TD>
      <TD className="font-mono-sm text-tabular text-right">16.7%</TD>
      <TD className="font-mono-sm text-tabular text-right">3,827</TD>
      <TD className="font-mono-sm text-tabular text-right">17.4%</TD>
      <TD className="font-mono-sm text-tabular text-right">3,381</TD>
    </tr>
    <tr>
      <TH rowHeader>Outlays</TH>
      <TD className="font-mono-sm text-tabular text-right">20.6%</TD>
      <TD className="font-mono-sm text-tabular text-right">3,982</TD>
      <TD className="font-mono-sm text-tabular text-right">20.2%</TD>
      <TD className="font-mono-sm text-tabular text-right">4,142</TD>
      <TD className="font-mono-sm text-tabular text-right">21.0%</TD>
      <TD className="font-mono-sm text-tabular text-right">4,470</TD>
      <TD className="font-mono-sm text-tabular text-right">21.3%</TD>
      <TD className="font-mono-sm text-tabular text-right">4,685</TD>
      <TD className="font-mono-sm text-tabular text-right">21.6%</TD>
      <TD className="font-mono-sm text-tabular text-right">4,949</TD>
      <TD className="font-mono-sm text-tabular text-right">20.3%</TD>
      <TD className="font-mono-sm text-tabular text-right">4,198</TD>
    </tr>
    <tr>
      <TH rowHeader>Budget Deficit</TH>
      <TD className="font-mono-sm text-tabular text-right">-3.5%</TD>
      <TD className="font-mono-sm text-tabular text-right">-665</TD>
      <TD className="font-mono-sm text-tabular text-right">-3.8%</TD>
      <TD className="font-mono-sm text-tabular text-right">-804</TD>
      <TD className="font-mono-sm text-tabular text-right">-4.6%</TD>
      <TD className="font-mono-sm text-tabular text-right">-981</TD>
      <TD className="font-mono-sm text-tabular text-right">-4.6%</TD>
      <TD className="font-mono-sm text-tabular text-right">-1,008</TD>
      <TD className="font-mono-sm text-tabular text-right">-4.9%</TD>
      <TD className="font-mono-sm text-tabular text-right">-1,123</TD>
      <TD className="font-mono-sm text-tabular text-right">-2.9%</TD>
      <TD className="font-mono-sm text-tabular text-right">-483</TD>
    </tr>
    <tr>
      <TH rowHeader>Debt Held by Public</TH>
      <TD className="font-mono-sm text-tabular text-right">76.0%</TD>
      <TD className="font-mono-sm text-tabular text-right">14,665</TD>
      <TD className="font-mono-sm text-tabular text-right">77.4%</TD>
      <TD className="font-mono-sm text-tabular text-right">15,688</TD>
      <TD className="font-mono-sm text-tabular text-right">79.2%</TD>
      <TD className="font-mono-sm text-tabular text-right">16,762</TD>
      <TD className="font-mono-sm text-tabular text-right">80.9%</TD>
      <TD className="font-mono-sm text-tabular text-right">17,872</TD>
      <TD className="font-mono-sm text-tabular text-right">83.1%</TD>
      <TD className="font-mono-sm text-tabular text-right">18,998</TD>
      <TD className="font-mono-sm text-tabular text-right">41.7%</TD>
      <TD className="font-mono-sm text-tabular text-right">8,065</TD>
    </tr>
  </tbody>,
];

export const StackedTableChildren = [
  <thead key="StackedTableChildren__thead">
    <tr>
      <TH>Document title</TH>
      <TH>Description</TH>
      <TH>Year</TH>
    </tr>
  </thead>,
  <tbody key="StackedTableChildren__tbody">
    <tr>
      <TH data-label="Document title" rowHeader>
        Declaration of Independence
      </TH>
      <TD data-label="Description">
        Statement adopted by the Continental Congress declaring independence
        from the British Empire.
      </TD>
      <TD data-label="Year">1776</TD>
    </tr>
    <tr>
      <TH data-label="Document title" rowHeader>
        Bill of Rights
      </TH>
      <TD data-label="Description">
        The first ten amendments of the U.S. Constitution guaranteeing rights
        and freedoms.
      </TD>
      <TD data-label="Year">1791</TD>
    </tr>
    <tr>
      <TH data-label="Document title" rowHeader>
        Declaration of Sentiments
      </TH>
      <TD data-label="Description">
        A document written during the Seneca Falls Convention outlining the
        rights that American women should be entitled to as citizens.
      </TD>
      <TD data-label="Year">1848</TD>
    </tr>
    <tr>
      <TH data-label="Document title" rowHeader>
        Emancipation Proclamation
      </TH>
      <TD data-label="Description">
        An executive order granting freedom to slaves in designated southern
        states.
      </TD>
      <TD data-label="Year">1863</TD>
    </tr>
  </tbody>,
];

export const SortableTableChildren = [
  <thead key="SortableTableChildren__thead">
    <tr>
      <TH dataSortable>Name</TH>
      <TH dataSortable sorted>
        Order admitted to union
      </TH>
      <TH dataSortable>Date of ratification vote</TH>
      <TH dataSortable>Date admitted to union</TH>
      <TH dataSortable>Percent of voters in favor of ratification</TH>
      <TH dataSortable>Votes cast in favor of ratification</TH>
      <TH dataSortable>Estimated population at time of admission</TH>
    </tr>
  </thead>,
  <tbody key="SortableTableChildren__tbody">
    <tr>
      <TH rowHeader>Hawaii</TH>
      <TD dataSortValue="50">50th</TD>
      <TD dataSortValue="331844400">Jun. 27, 1959</TD>
      <TD dataSortValue="327092400">Aug. 21, 1959</TD>
      <TD
        dataSortValue="0.943"
        className="font-mono-sm text-tabular text-right"
      >
        94.3%
      </TD>
      <TD
        dataSortValue="132773"
        className="font-mono-sm text-tabular text-right"
      >
        132,773
      </TD>
      <TD
        dataSortValue="632772"
        className="font-mono-sm text-tabular text-right"
      >
        632,772
      </TD>
    </tr>
    <tr>
      <TH rowHeader>Alaska</TH>
      <TD dataSortValue="49">49th</TD>
      <TD dataSortValue="431978400">Apr. 24, 1956</TD>
      <TD dataSortValue="346960800">Jan. 3, 1959</TD>
      <TD
        dataSortValue="0.681"
        className="font-mono-sm text-tabular text-right"
      >
        68.1%
      </TD>
      <TD
        dataSortValue="17477"
        className="font-mono-sm text-tabular text-right"
      >
        17,477
      </TD>
      <TD
        dataSortValue="226167"
        className="font-mono-sm text-tabular text-right"
      >
        226,167
      </TD>
    </tr>
    <tr>
      <TH rowHeader>Arizona</TH>
      <TD dataSortValue="48">48th</TD>
      <TD dataSortValue="1858528800">Feb. 9, 1911</TD>
      <TD dataSortValue="1826560800">Feb. 14, 1912</TD>
      <TD
        dataSortValue="0.787"
        className="font-mono-sm text-tabular text-right"
      >
        78.7%
      </TD>
      <TD
        dataSortValue="12187"
        className="font-mono-sm text-tabular text-right"
      >
        12,187
      </TD>
      <TD
        dataSortValue="204354"
        className="font-mono-sm text-tabular text-right"
      >
        204,354
      </TD>
    </tr>
    <tr>
      <TH rowHeader>New Mexico</TH>
      <TD dataSortValue="47">47th</TD>
      <TD dataSortValue="1835287200">Nov. 5, 1911</TD>
      <TD dataSortValue="1829930400">Jan. 6, 1912</TD>
      <TD
        dataSortValue="0.703"
        className="font-mono-sm text-tabular text-right"
      >
        70.3%
      </TD>
      <TD
        dataSortValue="31742"
        className="font-mono-sm text-tabular text-right"
      >
        31,742
      </TD>
      <TD
        dataSortValue="327301"
        className="font-mono-sm text-tabular text-right"
      >
        327,301
      </TD>
    </tr>
    <tr>
      <TH rowHeader>Oklahoma</TH>
      <TD dataSortValue="46">46th</TD>
      <TD dataSortValue="1965751200">Sep. 17, 1907</TD>
      <TD dataSortValue="1960567200">Nov. 16, 1907</TD>
      <TD
        dataSortValue="0.712"
        className="font-mono-sm text-tabular text-right"
      >
        71.2%
      </TD>
      <TD
        dataSortValue="180333"
        className="font-mono-sm text-tabular text-right"
      >
        180,333
      </TD>
      <TD
        dataSortValue="1657155"
        className="font-mono-sm text-tabular text-right"
      >
        1,657,155
      </TD>
    </tr>
    <tr>
      <TH rowHeader>Utah</TH>
      <TD dataSortValue="45">45th</TD>
      <TD dataSortValue="2340122400">Nov. 5, 1895</TD>
      <TD dataSortValue="2334938400">Jan. 4, 1896</TD>
      <TD
        dataSortValue="0.805"
        className="font-mono-sm text-tabular text-right"
      >
        80.5%
      </TD>
      <TD
        dataSortValue="31305"
        className="font-mono-sm text-tabular text-right"
      >
        31,305
      </TD>
      <TD
        dataSortValue="210779"
        className="font-mono-sm text-tabular text-right"
      >
        210,779
      </TD>
    </tr>
  </tbody>,
];
