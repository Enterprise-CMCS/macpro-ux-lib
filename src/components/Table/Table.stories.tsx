import React, { useDebugValue } from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Table } from "./Table";
import data from "./data.json";

export default {
  title: "COMPONENTS/Table",
  component: Table,
  args: {
    caption:
      "Standard Table: A commonly used table with many features and functionality",
    data: data["standardTable"],
  },
  argTypes: {},
} as ComponentMeta<typeof Table>;

const Template: ComponentStory<typeof Table> = ({ ...rest }) => (
  <Table {...rest} />
);

export const StandardTable = Template.bind({});
StandardTable.args = {};

export const BorderlessTable = Template.bind({});
BorderlessTable.args = {
  borderless: true,
  data: data["standardTable"],
  caption:
    "Borderless table: A borderless table can be useful when you want the information to feel more a part of the text it accompanies and extends.",
};

export const ScrollableTable = Template.bind({});
ScrollableTable.args = {
  data: data["scrollableTable"],
  footnote: "* in billions of dollars. Data for illustration purposes only.",
  scrollable: true,
};

export const SortableTable = Template.bind({});
SortableTable.args = {
  caption: "Recently admitted US states (sortable table example)",
  data: data["sortableTable"],
};
