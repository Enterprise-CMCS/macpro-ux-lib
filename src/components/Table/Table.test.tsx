import React from "react";
import { fireEvent, screen, render } from "../../test-setup";
import { Table } from "./Table";
import {
  BasicTableChildren,
  ScrollableTableChildren,
  StackedTableChildren,
  SortableTableChildren,
} from "./TableChildren";

describe("Tests for the Table component", () => {
  it("should render", () => {
    const { container } = render(<Table children={BasicTableChildren} />);

    // confirm basic structure of rendered table
    const rows = container.getElementsByTagName("tr");

    expect(rows.length).toBe(5);

    const colHeaders = rows[0];
    expect(colHeaders.children.length).toBe(3);

    expect(colHeaders.children[0].textContent).toBe("Document title");
    expect(colHeaders.children[1].textContent).toBe("Description");
    expect(colHeaders.children[2].textContent).toBe("Year");

    // confirm default classNames present
    expect(container.firstElementChild!.className).toBe("usa-table-container");

    const table = container.getElementsByTagName("table")[0];
    expect(table.className).toBe("usa-table");

    // caption should not exist
    const caption = container.getElementsByTagName("caption");
    expect(caption.length).toBe(0);

    // footnote should not exist
    const footnote = container.getElementsByClassName("usa-table__footnote");
    expect(footnote.length).toBe(0);
  });

  it("should be borderless, compact, scrollable, and striped", () => {
    const { container } = render(
      <Table
        children={BasicTableChildren}
        borderless
        compact
        scrollable
        striped
      />
    );

    expect(container.firstElementChild!.className).toBe(
      "usa-table-container usa-table-container--scrollable"
    );

    const table = container.getElementsByTagName("table")[0];
    expect(table.className).toBe(
      "usa-table usa-table--borderless usa-table--compact usa-table--striped"
    );
  });

  it("should be stacked", () => {
    const { container } = render(
      <Table children={BasicTableChildren} stacked />
    );

    expect(container.firstElementChild!.className).toBe(
      "usa-table-container width-mobile"
    );

    const table = container.getElementsByTagName("table")[0];
    expect(table.className).toBe("usa-table usa-table--stacked");
  });

  it("should be stacked with headers", () => {
    const { container } = render(
      <Table children={BasicTableChildren} stacked stackedHeader />
    );

    expect(container.firstElementChild!.className).toBe(
      "usa-table-container width-mobile"
    );

    const table = container.getElementsByTagName("table")[0];
    expect(table.className).toBe("usa-table usa-table--stacked-header");
  });

  it("should have a caption, a footnote, and custom className", () => {
    const { container } = render(
      <Table
        caption="This text should appear in the caption"
        children={BasicTableChildren}
        className="test-me"
        footnote="This text should appear in the footnote"
      />
    );

    // caption should exist
    const caption = container.getElementsByTagName("caption");
    expect(caption.length).toBe(1);
    expect(caption[0].innerHTML).toBe("This text should appear in the caption");

    // footnote should not exist
    const footnote = container.getElementsByClassName("usa-table__footnote");
    expect(footnote.length).toBe(1);
    expect(footnote[0].innerHTML).toBe(
      "This text should appear in the footnote"
    );

    // test custom class
    const table = container.getElementsByTagName("table")[0];
    expect(table.className).toBe("usa-table test-me");
  });
});

describe("snapshots for Table component", () => {
  it("Basic table", () => {
    const { container } = render(<Table children={BasicTableChildren} />);

    expect(container).toMatchSnapshot();
  });

  it("Scrollable table", () => {
    const { container } = render(
      <Table
        children={ScrollableTableChildren}
        caption="This text should appear in the caption"
        footnote="This text should appear in the footnote"
        scrollable
      />
    );

    expect(container).toMatchSnapshot();
  });
});
