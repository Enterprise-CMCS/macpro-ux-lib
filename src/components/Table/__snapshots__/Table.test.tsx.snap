// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots for Table component Basic table 1`] = `
<div>
  <div
    class="usa-table-container"
    tabindex="0"
  >
    <table
      class="usa-table"
    >
      <thead>
        <tr>
          <th
            role="columnheader"
            scope="col"
          >
            Document title
          </th>
          <th
            role="columnheader"
            scope="col"
          >
            Description
          </th>
          <th
            role="columnheader"
            scope="col"
          >
            Year
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Declaration of Independence
          </th>
          <td
            class=""
          >
            Statement adopted by the Continental Congress declaring independence from the British Empire.
          </td>
          <td
            class=""
          >
            1776
          </td>
        </tr>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Bill of Rights
          </th>
          <td
            class=""
          >
            The first ten amendments of the U.S. Constitution guaranteeing rights and freedoms.
          </td>
          <td
            class=""
          >
            1791
          </td>
        </tr>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Declaration of Sentiments
          </th>
          <td
            class=""
          >
            A document written during the Seneca Falls Convention outlining the rights that American women should be entitled to as citizens.
          </td>
          <td
            class=""
          >
            1848
          </td>
        </tr>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Emancipation Proclamation
          </th>
          <td
            class=""
          >
            An executive order granting freedom to slaves in designated southern states.
          </td>
          <td
            class=""
          >
            1863
          </td>
        </tr>
      </tbody>
    </table>
    <div
      aria-live="polite"
      class="usa-sr-only usa-table__announcement-region"
    />
  </div>
</div>
`;

exports[`snapshots for Table component Scrollable table 1`] = `
<div>
  <div
    class="usa-table-container usa-table-container--scrollable"
    tabindex="0"
  >
    <table
      class="usa-table"
    >
      <caption>
        This text should appear in the caption
      </caption>
      <colgroup>
        <col />
      </colgroup>
      <colgroup
        span="2"
      />
      <colgroup
        span="2"
      />
      <thead>
        <tr>
          <th
            role="columnheader"
            rowspan="2"
            scope="col"
          >
            Federal Budget
            <br />
            Baseline Projections
          </th>
          <th
            class="text-center"
            colspan="2"
            role="columnheader"
            scope="colgroup"
          >
            2017
          </th>
          <th
            class="text-center"
            colspan="2"
            role="columnheader"
            scope="colgroup"
          >
            2018
          </th>
          <th
            class="text-center"
            colspan="2"
            role="columnheader"
            scope="colgroup"
          >
            2019
          </th>
          <th
            class="text-center"
            colspan="2"
            role="columnheader"
            scope="colgroup"
          >
            2020
          </th>
          <th
            class="text-center"
            colspan="2"
            role="columnheader"
            scope="colgroup"
          >
            2021
          </th>
          <th
            class="text-center"
            colspan="2"
            role="columnheader"
            scope="colgroup"
          >
            Hist. Avg.
          </th>
        </tr>
        <tr>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            %GDP
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            USD*
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            %GDP
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            USD*
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            %GDP
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            USD*
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            %GDP
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            USD*
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            %GDP
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            USD*
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            %GDP
          </th>
          <th
            class="text-right"
            role="columnheader"
            scope="col"
          >
            USD*
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Revenue
          </th>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            17.2%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            3,316
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            16.4%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            3,338
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            16.3%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            3,490
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            16.7%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            3,678
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            16.7%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            3,827
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            17.4%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            3,381
          </td>
        </tr>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Outlays
          </th>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            20.6%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            3,982
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            20.2%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            4,142
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            21.0%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            4,470
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            21.3%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            4,685
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            21.6%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            4,949
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            20.3%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            4,198
          </td>
        </tr>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Budget Deficit
          </th>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -3.5%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -665
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -3.8%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -804
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -4.6%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -981
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -4.6%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -1,008
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -4.9%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -1,123
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -2.9%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            -483
          </td>
        </tr>
        <tr>
          <th
            role="rowheader"
            scope="row"
          >
            Debt Held by Public
          </th>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            76.0%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            14,665
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            77.4%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            15,688
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            79.2%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            16,762
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            80.9%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            17,872
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            83.1%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            18,998
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            41.7%
          </td>
          <td
            class="font-mono-sm text-tabular text-right"
          >
            8,065
          </td>
        </tr>
      </tbody>
    </table>
    <div
      aria-live="polite"
      class="usa-sr-only usa-table__announcement-region"
    />
    <p
      class="margin-bottom-3 usa-table__footnote"
    >
      This text should appear in the footnote
    </p>
  </div>
</div>
`;
